### PR TITLE
Added location router

### LINKED ISSUE #199: Location Service Endpoint
Currently, we are using Geodb Cities API, which is very incomplete. For example, Aveiro is not available.

We should implement our own endpoint to search for possible locations

### PR DESCRIPTION
**Pull request by Naapperas**
Closes #199

### CODE REVIEW

File: src/api/middleware/validators/location.js [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+
+import { useExpressValidators } from "../errorHandler.js";
+import ValidationReasons from "./validationReasons.js";
+
+import LocationConstants from "../../../models/constants/Location.js";
+
+export const search = useExpressValidators([
+    query("searchTerm", ValidationReasons.DEFAULT)
+        .isString().withMessage(ValidationReasons.STRING).bail()
+        .isLength({ min: 3 }).withMessage(ValidationReasons.LOCATION_SEARCH_TERM_MIN_LENGTH(LocationConstants.searchTerm.minLength)),
```
Comment (imnotteixeira) [inline src/api/middleware/validators/location.js:11]:
```suggestion
        .isLength({ min: LocationConstants.searchTerm.minLength }).withMessage(ValidationReasons.LOCATION_SEARCH_TERM_MIN_LENGTH(LocationConstants.searchTerm.minLength)),
```

Comment (imnotteixeira) [reply src/api/middleware/validators/location.js:11]: There is a `TOO_SHORT` validation reason that should be reused. THe error object will already tell you which field it is, no need to create more and more reasons.

File: src/loaders/mongoose.js [hunk 2]
```diff
@@ -10,25 +10,46 @@ export default async () => {
...
         await createDefaultAdmin();
     } catch (err) {
         console.error("Mongoose: failure in initial connection to the DB (aborting, will not retry)", err);
         process.exit(44);
     }
 
-    mongoose.connection.on("error", (err) => {
-        console.error("Mongoose connection error:", err);
+    mongoose.connections.forEach((connection) => {
+        connection.on("error", (err) => { // default mongoose connection used for the main app DB
```
Comment (imnotteixeira) [inline src/loaders/mongoose.js:50]: Is the comment accurate? Shouldn't this be iterating all the connections? Inc. the location one?

File: src/models/Location.js [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+    }
+});
+
+LocationSchema.index(
+    { citySearch: "text", countrySearch: "text" },
+    { name: "Search index", weights: { citySearch: 10, countrySearch: 5 } }
+);
+
+// this works because mongoose caches commands until a valid connection can be established or a timeout error is raised
+const Location = mongoose.connection.useDb(config.location_db_name).model("Location", LocationSchema);
```
Comment (imnotteixeira) [inline src/models/Location.js:40]: We need to pass a `{useCache: true}` to the useDb options

**Review by imnotteixeira — CHANGES_REQUESTED** [review]

File: src/loaders/mongoose.js (hunk 2 above)
Comment (Naapperas) [reply src/loaders/mongoose.js:50]: I just copied it over, forgot to change/remove it

Comment (Naapperas) [reply src/loaders/mongoose.js:50]: Also, before this I explicitely called the on("error") on both the connection and connections[1] objects, so this is in fact iterating both connections.

**Comment by Naapperas** [comment]
After discussing some issues regarding the implementations of these features, it has been concluded that this behavior is best achieved in another application instead of being built into NIJobs, so this PR will be closed after those features have been implemented elsewhere.

**Pull request closed without merging**